### ISSUE TITLE
Treesitter language injections can enter infinite loop

### ISSUE THREAD

**Issue by ii14** [neovim/neovim#21275]
### Neovim version (nvim -v)

v0.9.0-dev, 1f6d36c2c

### Vim (not Nvim) behaves the same?

-

### Operating system/version

ubuntu18.04

### Terminal name/version

alacritty

### $TERM environment variable

alacritty

### Installation

unstable ppa

### How to reproduce the issue

`nvim --clean`, run the script:
```lua
vim.treesitter.get_string_parser([[
#define TEST \
  0
]], 'c'):parse()
```

### Expected behavior

don't enter infinite loop

### Actual behavior

Enters infinite loop here: https://github.com/neovim/neovim/blob/fba0562723a1af143c9e9509920d03d8231b8bf7/runtime/lua/vim/treesitter/languagetree.lua#L166

**Comment by clason** [comment]
Probably due to the self-injections for macros:
https://github.com/neovim/neovim/blob/fba0562723a1af143c9e9509920d03d8231b8bf7/runtime/queries/c/injections.scm#L1

Can you try removing that file (or putting an empty `injections.scm` in `stdpath('config')/queries/c`?

**Comment by ii14** [comment]
Yes, creating empty `injections.scm` works.

**Comment by clason** [comment]
(You can -- and for string parsers should -- override the queries used with `vim.treesitter.set_query()`.)

**Comment by clason** [comment]
Of course, that doesn't mean we shouldn't add a timeout/recursion limit to the parsing.

**Comment by justinmk** [comment]
With the repro steps on master 9b25c68db21c , I don't see an error nor an infinite loop. 

Presumably fixed by https://github.com/neovim/neovim/pull/32468
